Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -0,0 +1,333 @@
+ * `WeierstrassCurve.Jacobian.nonsingular_lift`: the nonsingular condition on a point class.
+
+## Implementation notes
+
+A point representative is implemented as a term `P` of type `Fin 3 → R`, which allows for the vector
+notation `![x, y, z]`. However, `P` is not definitionally equivalent to the expanded vector
+`![P x, P y, P z]`, so the auxiliary lemma `fin3_def` can be used to convert between the two forms.
+The equivalence of two point representatives `P` and `Q` is implemented as an equivalence of orbits
+of the action of `Rˣ`, or equivalently that there is some unit `u` of `R` such that `P = u • Q`.
+However, `u • Q` is again not definitionally equal to `![u² * Q x, u³ * Q y, u * Q z]`, so the
+auxiliary lemmas `smul_fin3` and `smul_fin3_ext` can be used to convert between the two forms.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, rational point, Jacobian coordinates
+-/
+
+local notation "x" => 0
+
+local notation "y" => 1
+
+local notation "z" => 2
+
+local macro "matrix_simp" : tactic =>
+  `(tactic| simp only [Matrix.head_cons, Matrix.tail_cons, Matrix.smul_empty, Matrix.smul_cons,
+    Matrix.cons_val_zero, Matrix.cons_val_one, Matrix.cons_val_two])
+
+universe u
+
+/-! ## Weierstrass curves -/
+
+/-- An abbreviation for a Weierstrass curve in Jacobian coordinates. -/
+abbrev WeierstrassCurve.Jacobian :=
+  WeierstrassCurve
+
+namespace WeierstrassCurve.Jacobian
+
+open MvPolynomial
+
+local macro "eval_simp" : tactic =>
+  `(tactic| simp only [eval_C, eval_X, eval_add, eval_sub, eval_mul, eval_pow])
+
+local macro "pderiv_simp" : tactic =>
+  `(tactic| simp only [map_ofNat, map_neg, map_add, map_sub, map_mul, pderiv_mul, pderiv_pow,
+    pderiv_C, pderiv_X_self, pderiv_X_of_ne one_ne_zero, pderiv_X_of_ne one_ne_zero.symm,
+    pderiv_X_of_ne (by decide : (2 : Fin 3) ≠ 0), pderiv_X_of_ne (by decide : (0 : Fin 3) ≠ 2),
+    pderiv_X_of_ne (by decide : (2 : Fin 3) ≠ 1), pderiv_X_of_ne (by decide : (1 : Fin 3) ≠ 2)])
+
+variable {R : Type u} [CommRing R] (W : Jacobian R)
+
+lemma fin3_def (P : Fin 3 → R) : P = ![P x, P y, P z] := by
+  ext n; fin_cases n <;> rfl
+
+/-- The scalar multiplication on a point representative. -/
+scoped instance instSMulPoint : SMul Rˣ <| Fin 3 → R :=
+  ⟨fun u P => ![u ^ 2 * P x, u ^ 3 * P y, u * P z]⟩
+
+lemma smul_fin3 (P : Fin 3 → R) (u : Rˣ) :
+    u • P = ![u ^ 2 * P x, u ^ 3 * P y, u * P z] :=
+  rfl
+
+lemma smul_fin3_ext (P : Fin 3 → R) (u : Rˣ) :
+    (u • P) x = u ^ 2 * P x ∧ (u • P) y = u ^ 3 * P y ∧ (u • P) z = u * P z :=
+  ⟨rfl, rfl, rfl⟩
+
+/-- The multiplicative action on a point representative. -/
+scoped instance instMulActionPoint : MulAction Rˣ <| Fin 3 → R where
+  one_smul := fun _ => by
+    simp only [smul_fin3, Units.val_one, one_pow, one_mul, ← fin3_def]
+  mul_smul := fun u v P => by
+    simp only [smul_fin3, Units.val_mul, mul_pow, mul_assoc]
+    matrix_simp
```

Comment:
```suggestion
  one_smul _ := by simp only [smul_fin3, Units.val_one, one_pow, one_mul, ← fin3_def]
  mul_smul u v P := by simp only [smul_fin3, Units.val_mul, mul_pow, mul_assoc]; matrix_simp
```
(untested)

Reply:
Yes, thanks! This is not too important right now, but in the future I plan to combine all the macros common to the `EllipticCurve` files, and will see what I can get out of this.